pkg/api/filters/filter.go

```diff
@@ -0,0 +1,40 @@
+package filters
+
+import "net/http"
+
+//Filter is a wrapper around a http.HandlerFunc.
+type Filter interface {
+	Execute(http.HandlerFunc) http.HandlerFunc
+}
+
+//Chain represents a chain of Filters that can be applied
+//to a http.HandlrFunc.
+type Chain interface {
+	Filter(http.HandlerFunc) http.HandlerFunc
+}
+
+//Chain contains a number of HandlerFunctions
+type chain struct {
+	filters []Filter
+}
+
+//NewFilterChain returns a chain struct populated with
+//the Filter(s) provided
+func NewFilterChain(filterChain []Filter) Chain {
+	return chain{
+		filters: filterChain,
+	}
+}
+
+//Filter applies a chain of FilterFunction to a target http.HandlerFun
+//and returns a new http.HandlerFunc that wraps the chain
+func (h chain) Filter(
```

Comment:
The heart of this change is contained in this file. Up above, two new interfaces are defined. A Filter, which can execute on a http.HandlerFunc and return another one. The Chain interface defines a Filter method. The Chain is implemented by the chain struct, which has a slice of Filter(s). When it's Filter method is called with a http.HandlerFunc (i.e. 	filterChain.Filter(s.provision)), it wraps that argument with the filter slice in reverse order.

Reply:
Q: Do we really need a separate interface for `FilterChain`? Cannot `FilterChain` be a specific _implementation_ of the `Filter` interface that just has the logic to chain a composition of _other_ `Filter` implementations together?

Reply:
I don't think we do, given they are the same. I think that these can be consolidated and FilterChain is just another filter, like you mentioned.

Reply:
There might be a very good reason for it, btw. You tell me.

Reply:
I don't think so. It's mainly an artifact of me working through it in pieces. I think conceptually it's the same thing. It's a "filter" that applies other "filters" in the end,